File: Fundamentals_level_5/Localmind/frontend/src/lib/api/sources.ts
/**
 * Sources API Service
 * Educational Note: Handles all source operations with the backend.
 * Sources are documents, images, audio, and data files uploaded to projects.
 */

import axios from 'axios';

const API_BASE_URL = 'http://localhost:5000/api/v1';

/**
 * Source metadata returned from the API
 * Educational Note: Status transitions for sources:
 * - uploaded: File received, waiting for processing
 * - processing: Currently extracting text from PDF
 * - embedding: Creating vector embeddings for semantic search
 * - ready: Successfully processed, available for chat context
 * - error: Processing failed (no partial states - clean failure)
 */
export interface Source {
  id: string;
  project_id: string;
  name: string;
  original_filename: string;
  description: string;
  category: 'document' | 'audio' | 'image' | 'data' | 'link';
  mime_type: string;
  file_extension: string;
  file_size: number;
  stored_filename: string;
  status: 'uploaded' | 'processing' | 'embedding' | 'ready' | 'error';
  active: boolean; // Whether source is included in chat context
  processing_info: Record<string, unknown> | null;
  embedding_info?: Record<string, unknown> | null; // Embedding details
  created_at: string;
  updated_at: string;
}

/**
 * Summary statistics for sources
 */
export interface SourcesSummary {
  total_count: number;
  total_size: number;
  by_category: Record<string, number>;
  by_status: Record<string, number>;
}

/**
 * Allowed file extensions grouped by category
 */
export interface AllowedTypes {
  allowed_extensions: Record<string, string>;
  by_category: Record<string, string[]>;
}

// Maximum number of sources allowed per project
export const MAX_SOURCES = 100;

// Maximum image file size (5MB) - API constraint
export const MAX_IMAGE_SIZE = 5 * 1024 * 1024;

// Allowed file extensions by category
export const ALLOWED_EXTENSIONS = {
  document: ['.pdf', '.txt', '.docx', '.pptx'],
  image: ['.jpeg', '.jpg', '.png', '.gif', '.webp'],
  audio: ['.mp3', '.wav', '.m4a', '.aac', '.flac'],
  data: ['.csv'],
};

class SourcesAPI {
  /**
   * List all sources for a project
   */
  async listSources(projectId: string): Promise<Source[]> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/sources`
      );
      return response.data.sources;
    } catch (error) {
      console.error('Error fetching sources:', error);
      throw error;
    }
  }

  /**
   * Upload a new source file
   * Educational Note: Uses FormData for multipart file upload
   */
  async uploadSource(
    projectId: string,
    file: File,
    name?: string,
    description?: string
  ): Promise<Source> {
    try {
      const formData = new FormData();
      formData.append('file', file);
      if (name) formData.append('name', name);
      if (description) formData.append('description', description);

      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/sources`,
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        }
      );
      return response.data.source;
    } catch (error) {
      console.error('Error uploading source:', error);
      throw error;
    }
  }

  /**
   * Get a specific source's metadata
   */
  async getSource(projectId: string, sourceId: string): Promise<Source> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/sources/${sourceId}`
      );
      return response.data.source;
    } catch (error) {
      console.error('Error fetching source:', error);
      throw error;
    }
  }

  /**
   * Update a source's metadata
   */
  async updateSource(
    projectId: string,
    sourceId: string,
    data: { name?: string; description?: string; active?: boolean }
  ): Promise<Source> {
    try {
      const response = await axios.put(
        `${API_BASE_URL}/projects/${projectId}/sources/${sourceId}`,
        data
      );
      return response.data.source;
    } catch (error) {
      console.error('Error updating source:', error);
      throw error;
    }
  }

  /**
   * Delete a source
   */
  async deleteSource(projectId: string, sourceId: string): Promise<void> {
    try {
      await axios.delete(
        `${API_BASE_URL}/projects/${projectId}/sources/${sourceId}`
      );
    } catch (error) {
      console.error('Error deleting source:', error);
      throw error;
    }
  }

  /**
   * Get the download URL for a source
   */
  getDownloadUrl(projectId: string, sourceId: string): string {
    return `${API_BASE_URL}/projects/${projectId}/sources/${sourceId}/download`;
  }

  /**
   * Get sources summary (counts and sizes)
   */
  async getSourcesSummary(projectId: string): Promise<SourcesSummary> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/sources/summary`
      );
      return response.data.summary;
    } catch (error) {
      console.error('Error fetching sources summary:', error);
      throw error;
    }
  }

  /**
   * Get allowed file types for upload
   */
  async getAllowedTypes(): Promise<AllowedTypes> {
    try {
      const response = await axios.get(`${API_BASE_URL}/sources/allowed-types`);
      return {
        allowed_extensions: response.data.allowed_extensions,
        by_category: response.data.by_category,
      };
    } catch (error) {
      console.error('Error fetching allowed types:', error);
      throw error;
    }
  }

  /**
   * Add a URL source (website or YouTube link)
   * Educational Note: URLs are stored as .link files containing JSON metadata.
   * The actual content fetching happens in a separate processing step.
   */
  async addUrlSource(
    projectId: string,
    url: string,
    name?: string,
    description?: string
  ): Promise<Source> {
    try {
      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/sources/url`,
        { url, name, description }
      );
      return response.data.source;
    } catch (error) {
      console.error('Error adding URL source:', error);
      throw error;
    }
  }

  /**
   * Add a pasted text source
   * Educational Note: Text is stored as a .txt file. This is the simplest
   * source type - the raw content IS the processed content.
   */
  async addTextSource(
    projectId: string,
    content: string,
    name: string,
    description?: string
  ): Promise<Source> {
    try {
      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/sources/text`,
        { content, name, description }
      );
      return response.data.source;
    } catch (error) {
      console.error('Error adding text source:', error);
      throw error;
    }
  }

  /**
   * Cancel processing for a source
   * Educational Note: This stops any running tasks and sets status back to "uploaded"
   * so user can retry later. Raw file is preserved, only processed data is deleted.
   */
  async cancelProcessing(projectId: string, sourceId: string): Promise<void> {
    try {
      await axios.post(
        `${API_BASE_URL}/projects/${projectId}/sources/${sourceId}/cancel`
      );
    } catch (error) {
      console.error('Error cancelling source processing:', error);
      throw error;
    }
  }

  /**
   * Retry processing for a failed or uploaded source
   * Educational Note: This restarts processing from the raw file.
   */
  async retryProcessing(projectId: string, sourceId: string): Promise<void> {
    try {
      await axios.post(
        `${API_BASE_URL}/projects/${projectId}/sources/${sourceId}/retry`
      );
    } catch (error) {
      console.error('Error retrying source processing:', error);
      throw error;
    }
  }
}

export const sourcesAPI = new SourcesAPI();

/**
 * Helper function to format file size
 */
export function formatFileSize(bytes: number): string {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(1)) + ' ' + sizes[i];
}

/**
 * Helper function to get icon name based on category
 */
export function getCategoryIcon(category: string): string {
  switch (category) {
    case 'document':
      return 'FileText';
    case 'audio':
      return 'Music';
    case 'image':
      return 'Image';
    case 'data':
      return 'Table';
    case 'link':
      return 'Link';
    default:
      return 'File';
  }
}

File: Fundamentals_level_5/Localmind/backend/app/api/sources.py
"""
API endpoints for managing project sources.

Educational Note: These endpoints handle uploading, listing, and managing
raw source files for projects. Sources are documents, images, audio, and
data files that provide context for AI conversations.
"""
from flask import jsonify, request, current_app, send_file
from app.api import api_bp
from app.services.source_service import SourceService

# Initialize service
source_service = SourceService()


@api_bp.route('/projects/<project_id>/sources', methods=['GET'])
def list_sources(project_id: str):
    """
    List all sources for a project.

    Educational Note: Returns metadata for all uploaded sources,
    sorted by most recent first.
    """
    try:
        sources = source_service.list_sources(project_id)

        return jsonify({
            'success': True,
            'sources': sources,
            'count': len(sources)
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error listing sources: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@api_bp.route('/projects/<project_id>/sources', methods=['POST'])
def upload_source(project_id: str):
    """
    Upload a new source file to a project.

    Educational Note: Accepts multipart/form-data with:
    - file: The source file (required)
    - name: Display name (optional, defaults to filename)
    - description: Description of the source (optional)

    Supported file types: PDF, TXT, MD, MP3, images (PNG, JPG, etc.), CSV
    """
    try:
        # Check if file is in request
        if 'file' not in request.files:
            return jsonify({
                'success': False,
                'error': 'No file provided'
            }), 400

        file = request.files['file']

        if not file.filename:
            return jsonify({
                'success': False,
                'error': 'No file selected'
            }), 400

        # Get optional fields from form data
        name = request.form.get('name')
        description = request.form.get('description', '')

        # Upload the source
        source = source_service.upload_source(
            project_id=project_id,
            file=file,
            name=name,
            description=description
        )

        return jsonify({
            'success': True,
            'source': source,
            'message': 'Source uploaded successfully'
        }), 201

    except ValueError as e:
        # Validation errors (file type not allowed, etc.)
        return jsonify({
            'success': False,
            'error': str(e)
        }), 400

    except Exception as e:
        current_app.logger.error(f"Error uploading source: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@api_bp.route('/projects/<project_id>/sources/<source_id>', methods=['GET'])
def get_source(project_id: str, source_id: str):
    """
    Get a specific source's metadata.

    Educational Note: Returns full metadata for a single source,
    including processing status and details.
    """
    try:
        source = source_service.get_source(project_id, source_id)

        if not source:
            return jsonify({
                'success': False,
                'error': 'Source not found'
            }), 404

        return jsonify({
            'success': True,
            'source': source
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error getting source: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@api_bp.route('/projects/<project_id>/sources/<source_id>', methods=['PUT'])
def update_source(project_id: str, source_id: str):
    """
    Update a source's metadata.

    Educational Note: Allows updating display name and description.
    The raw file cannot be updated - delete and re-upload instead.
    """
    try:
        data = request.get_json()

        if not data:
            return jsonify({
                'success': False,
                'error': 'No data provided'
            }), 400

        source = source_service.update_source(
            project_id=project_id,
            source_id=source_id,
            name=data.get('name'),
            description=data.get('description')
        )

        if not source:
            return jsonify({
                'success': False,
                'error': 'Source not found'
            }), 404

        return jsonify({
            'success': True,
            'source': source,
            'message': 'Source updated successfully'
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error updating source: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@api_bp.route('/projects/<project_id>/sources/<source_id>', methods=['DELETE'])
def delete_source(project_id: str, source_id: str):
    """
    Delete a source and its raw file.

    Educational Note: This is a hard delete - the raw file
    and all metadata are permanently removed.
    """
    try:
        deleted = source_service.delete_source(project_id, source_id)

        if not deleted:
            return jsonify({
                'success': False,
                'error': 'Source not found'
            }), 404

        return jsonify({
            'success': True,
            'message': 'Source deleted successfully'
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error deleting source: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@api_bp.route('/projects/<project_id>/sources/<source_id>/download', methods=['GET'])
def download_source(project_id: str, source_id: str):
    """
    Download the raw source file.

    Educational Note: Returns the original uploaded file
    for download or viewing.
    """
    try:
        source = source_service.get_source(project_id, source_id)

        if not source:
            return jsonify({
                'success': False,
                'error': 'Source not found'
            }), 404

        file_path = source_service.get_source_file_path(project_id, source_id)

        if not file_path or not file_path.exists():
            return jsonify({
                'success': False,
                'error': 'Source file not found on disk'
            }), 404

        # Send file with appropriate MIME type
        return send_file(
            file_path,
            mimetype=source.get('mime_type', 'application/octet-stream'),
            as_attachment=True,
            download_name=source.get('original_filename', source.get('name'))
        )

    except Exception as e:
        current_app.logger.error(f"Error downloading source: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@api_bp.route('/projects/<project_id>/sources/summary', methods=['GET'])
def get_sources_summary(project_id: str):
    """
    Get a summary of sources for a project.

    Educational Note: Returns aggregate statistics about sources,
    useful for dashboard displays.
    """
    try:
        summary = source_service.get_sources_summary(project_id)

        return jsonify({
            'success': True,
            'summary': summary
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error getting sources summary: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@api_bp.route('/sources/allowed-types', methods=['GET'])
def get_allowed_types():
    """
    Get the list of allowed file types for upload.

    Educational Note: Returns allowed extensions grouped by category.
    Useful for frontend validation and UI display.
    """
    try:
        extensions = source_service.get_allowed_extensions()

        # Group by category for easier frontend use
        by_category = {}
        for ext, category in extensions.items():
            if category not in by_category:
                by_category[category] = []
            by_category[category].append(ext)

        return jsonify({
            'success': True,
            'allowed_extensions': extensions,
            'by_category': by_category
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error getting allowed types: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@api_bp.route('/projects/<project_id>/sources/url', methods=['POST'])
def add_url_source(project_id: str):
    """
    Add a URL source (website or YouTube link) to a project.

    Educational Note: Accepts JSON with:
    - url: The URL to store (required)
    - name: Display name (optional, defaults to URL)
    - description: Description of the source (optional)

    The URL is stored as a .link file. Content fetching happens separately.
    """
    try:
        data = request.get_json()

        if not data:
            return jsonify({
                'success': False,
                'error': 'No data provided'
            }), 400

        url = data.get('url')
        if not url:
            return jsonify({
                'success': False,
                'error': 'URL is required'
            }), 400

        source = source_service.add_url_source(
            project_id=project_id,
            url=url,
            name=data.get('name'),
            description=data.get('description', '')
        )

        return jsonify({
            'success': True,
            'source': source,
            'message': 'URL source added successfully'
        }), 201

    except ValueError as e:
        # Validation errors (invalid URL, etc.)
        return jsonify({
            'success': False,
            'error': str(e)
        }), 400

    except Exception as e:
        current_app.logger.error(f"Error adding URL source: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@api_bp.route('/projects/<project_id>/sources/text', methods=['POST'])
def add_text_source(project_id: str):
    """
    Add a pasted text source to a project.

    Educational Note: Accepts JSON with:
    - content: The pasted text content (required)
    - name: Display name for the source (required)
    - description: Description of the source (optional)

    The text is stored as a .txt file.
    """
    try:
        data = request.get_json()

        if not data:
            return jsonify({
                'success': False,
                'error': 'No data provided'
            }), 400

        content = data.get('content')
        name = data.get('name')

        if not content:
            return jsonify({
                'success': False,
                'error': 'Content is required'
            }), 400

        if not name:
            return jsonify({
                'success': False,
                'error': 'Name is required'
            }), 400

        source = source_service.add_text_source(
            project_id=project_id,
            content=content,
            name=name,
            description=data.get('description', '')
        )

        return jsonify({
            'success': True,
            'source': source,
            'message': 'Text source added successfully'
        }), 201

    except ValueError as e:
        # Validation errors (empty content, etc.)
        return jsonify({
            'success': False,
            'error': str(e)
        }), 400

    except Exception as e:
        current_app.logger.error(f"Error adding text source: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500
